src/layer.rs

```diff
@@ -16,34 +16,197 @@ use crate::{Error, Glyph, Plist};
+    layers: Vec<Layer>,
+}
+
+#[allow(clippy::clippy::len_without_is_empty)] // never empty
+impl LayerSet {
+    /// Load the layers from the provided path.
+    ///
+    /// If a `layercontents.plist` file exists, it will be used, otherwise
+    /// we will assume the pre-UFOv3 behaviour, and expect a single glyphs dir.
+    ///
+    /// The `glyph_names` argument allows norad to reuse glyph name strings,
+    /// reducing memory use.
+    pub fn load(base_dir: &Path, glyph_names: &NameList) -> Result<LayerSet, Error> {
+        let layer_contents_path = base_dir.join(LAYER_CONTENTS_FILE);
+        let to_load: Vec<(LayerName, PathBuf)> = if layer_contents_path.exists() {
+            plist::from_file(layer_contents_path)?
+        } else {
+            vec![(Arc::from(DEFAULT_LAYER_NAME), PathBuf::from(DEFAULT_GLYPHS_DIRNAME))]
+        };
+
+        let mut layers: Vec<_> = to_load
+            .into_iter()
+            .map(|(name, path)| {
+                let layer_path = base_dir.join(&path);
+                Layer::load_impl(&layer_path, name, &glyph_names)
+            })
+            .collect::<Result<_, _>>()?;
+
+        // move the default layer to the front
+        let default_idx = layers
+            .iter()
+            .position(|l| l.path.to_str() == Some(DEFAULT_GLYPHS_DIRNAME))
+            .ok_or(Error::MissingDefaultLayer)?;
+        layers.rotate_left(default_idx);
+
+        Ok(LayerSet { layers })
+    }
+
+    /// The number of layers in the set.
+    ///
+    /// This should be non-zero.
+    pub fn len(&self) -> usize {
+        self.layers.len()
+    }
+
+    /// Get a reference to a layer, by name.
+    pub fn get(&self, name: &str) -> Option<&Layer> {
+        self.layers.iter().find(|l| &*l.name == name)
+    }
+
+    /// Get a mutable reference to a layer, by name.
+    pub fn get_mut(&mut self, name: &str) -> Option<&mut Layer> {
+        self.layers.iter_mut().find(|l| &*l.name == name)
+    }
+
+    /// A reference to the default layer.
+    pub fn default_layer(&self) -> &Layer {
+        &self.layers[0]
+    }
+
+    /// A mutable reference to the default layer.
+    pub fn default_layer_mut(&mut self) -> &mut Layer {
+        &mut self.layers[0]
+    }
+
+    /// Iterate over all layers.
+    pub fn iter(&self) -> impl Iterator<Item = &Layer> {
+        self.layers.iter()
+    }
+
+    /// Iterate over the names of all layers.
+    pub fn names(&self) -> impl Iterator<Item = &LayerName> {
+        self.layers.iter().map(|l| &l.name)
+    }
+
+    /// Create a new layer with the given name.
+    pub fn new_layer(&mut self, name: &str) -> Result<(), Error> {
+        if self.layers.iter().any(|l| &*l.name == name) {
+            Err(Error::DuplicateLayer(name.into()))
+        } else {
+            let layer = Layer::new(name.into(), None);
+            self.layers.push(layer);
+            Ok(())
+        }
+    }
+
+    /// Remove a layer.
+    ///
+    /// The default layer cannot be removed.
```

Comment:
I'm not seeing anything stopping client code from doing it anyway?

Reply:
Maybe needs a `iter().skip(1)` or something?

Reply:
ah good catch, I changed the implementation (we used to track the default layer name instead of just insisting that it is at the front) and this wasn't updated appropriately. 👍